src/group/mod.rs

```diff
@@ -200,3 +218,88 @@ impl<'a> Groups {
+
+    fn next_delegate_and_bcast_targets(group: &Groups) -> (u64, Vec<u64>) {
+        group
+            .bcast_targets
+            .iter()
+            .next()
+            .map(|(k, v)| (*k, v.clone()))
+            .unwrap()
+    }
+
+    #[test]
+    fn test_group() {
+        let mut group = Groups::new(vec![(1, vec![1, 2]), (2, vec![3, 4, 5]), (3, vec![6])]);
+        assert_eq!(group.unresolved.len(), 6);
+        group.set_leader_group_id(1);
+
+        let mut prs = ProgressSet::new(crate::default_logger());
+        for id in 1..=6 {
+            let mut pr = Progress::new(100, 100);
+            pr.matched = 99;
+            prs.insert_voter(id, pr).unwrap();
+        }
+
+        // After the resolving, only group 2 should be delegated.
+        group.resolve_delegates(&prs);
+        assert_eq!(group.bcast_targets.len(), 1);
+        let (delegate, mut targets) = next_delegate_and_bcast_targets(&group);
+        targets.push(delegate);
+        targets.sort();
+        assert_eq!(targets, [3, 4, 5]);
+
+        // Remove a delegate which doesn't exists.
+        group.remove_delegate(6);
+        assert!(group.unresolved.is_empty());
+
+        // Remove a peer which is not delegate.
+        let remove = match delegate {
+            3 => 4,
+            4 => 5,
+            5 => 3,
+            _ => unreachable!(),
+        };
+        group.remove_delegate(remove);
+        assert!(group.unresolved.is_empty());
+        let (_, targets) = next_delegate_and_bcast_targets(&group);
+        assert_eq!(targets.len(), 2);
+
+        // Remove a delegate.
+        group.remove_delegate(delegate);
+        assert!(!group.unresolved.is_empty());
+        group.resolve_delegates(&prs);
+        let (_, targets) = next_delegate_and_bcast_targets(&group);
+        assert_eq!(targets.len(), 1);
+
+        // Add the removed peer back, without group id.
+        let peer = delegate;
+        group.update_group_id(peer, INVALID_ID);
+        assert!(group.unresolved.is_empty());
+
+        // Add the removed peer back, with group id.
+        assert!(!group.update_group_id(peer, 2));
+        let (_, targets) = next_delegate_and_bcast_targets(&group);
+        assert_eq!(targets.len(), 2);
+
+        // The peer reports to the group again, without group id.
+        let (delegate, _) = next_delegate_and_bcast_targets(&group);
+        assert_ne!(peer, delegate);
+        group.update_group_id(peer, INVALID_ID);
+        let (_, targets) = next_delegate_and_bcast_targets(&group);
+        assert_eq!(targets.len(), 1);
+
+        // The delegate changes group to 3.
+        assert!(group.update_group_id(delegate, 3));
+        assert!(group.bcast_targets.is_empty());
+        group.resolve_delegates(&prs);
+        let (_, targets) = next_delegate_and_bcast_targets(&group);
+        assert!(targets.contains(&delegate) || targets.contains(&6));
+        assert!(group.get_delegate(6) == 6 || group.get_delegate(6) == delegate);
```

Comment:
L302 and L303 are really tricky 😂

Reply:
Seems no better way. Which peer is deleted is not certain.